lxd/storage/drivers/driver_dir.go

```diff
@@ -0,0 +1,864 @@
+func (d *dir) deleteQuota(path string, volID int64) error {
+	if volID == 0 {
+		return fmt.Errorf("Missing volume ID")
+	}
+
+	ok, err := quota.Supported(path)
+	if err != nil || !ok {
+		// Skipping quota as underlying filesystem doesn't suppport project quotas.
+		return nil
+	}
+
+	err = quota.SetProject(path, 0)
+	if err != nil {
+		return err
+	}
+
+	err = quota.SetProjectQuota(path, d.quotaProjectID(volID), 0)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// CreateVolumeSnapshot creates a snapshot of a volume.
+func (d *dir) CreateVolumeSnapshot(volType VolumeType, volName string, newSnapshotName string, op *operations.Operation) error {
+	if shared.IsSnapshot(volName) {
+		return fmt.Errorf("Volume cannot be snapshot")
+	}
+
+	if shared.IsSnapshot(newSnapshotName) {
+		return fmt.Errorf("Snapshot name is not a valid snapshot name")
+	}
+
+	// Get the volume ID for the parent volume, which is used to set project quota on snapshot.
+	volID, err := d.getVolID(volType, volName)
+	if err != nil {
+		return err
+	}
+
+	srcPath := GetVolumeMountPoint(d.name, volType, volName)
+	fullSnapName := fmt.Sprintf("%s%s%s", volName, shared.SnapshotDelimiter, newSnapshotName)
+	snapPath := GetVolumeMountPoint(d.name, volType, fullSnapName)
+
+	err = os.MkdirAll(snapPath, 0711)
+	if err != nil {
+		return err
+	}
+
+	revertPath := true
+	defer func() {
+		if revertPath {
+			d.deleteQuota(snapPath, volID)
+			os.RemoveAll(snapPath)
+		}
+	}()
+
+	// Initialise the snapshot's quota with the parent volume's ID.
+	err = d.initQuota(snapPath, volID)
+	if err != nil {
+		return err
+	}
+
+	// Not really needed, but useful to show how it should work for other drivers, also
+	// works the same as copying snapshots CreateVolumeFromCopy above.
+	ourMount, err := d.MountVolume(volType, fullSnapName, op)
+	if err != nil {
+		return err
+	}
+
+	bwlimit := d.config["rsync.bwlimit"]
+
+	_, err = rsync.LocalCopy(srcPath, snapPath, bwlimit, true)
+	if err != nil {
+		if ourMount {
+			d.UnmountVolume(volType, fullSnapName, op)
+		}
+		return err
+	}
+
+	if ourMount {
+		d.UnmountVolume(volType, fullSnapName, op)
+	}
+
+	revertPath = false
+	return nil
+}
+
+// DeleteVolumeSnapshot removes a snapshot from the storage device. The volName and snapshotName
+// must be bare names and should not be in the format "volume/snapshot".
+func (d *dir) DeleteVolumeSnapshot(volType VolumeType, volName string, snapshotName string, op *operations.Operation) error {
+	if shared.IsSnapshot(volName) {
+		return fmt.Errorf("Volume name must not be a snapshot")
+	}
+
+	if shared.IsSnapshot(snapshotName) {
+		return fmt.Errorf("Snapshot name must not contain volume name")
+	}
+
+	// Get the volume ID for the parent volume, which is used to remove project quota.
+	volID, err := d.getVolID(volType, volName)
+	if err != nil {
+		return err
+	}
+
+	snapPath := GetVolumeMountPoint(d.name, volType, fmt.Sprintf("%s%s%s", volName, shared.SnapshotDelimiter, snapshotName))
```

Comment:
I see this pattern going on quite a bit, should we have a GetVolumeSnapshotMountPoint function which does this?

Reply:
@stgraber agreed, have added a `GetSnapshotVolumeName(parent, snapName)` function that joins them to the snapshot separator, so it can be used in other places (like db row generation) rather than just mount path generation. I'm also keen to not repeat the logic used in  GetVolumeMountPoint if possible. But this new function will make things simpler to generate a snapshot mount path with GetVolumeMountPoint.